source/opt/fold.cpp

```diff
@@ -180,6 +182,53 @@ uint32_t OperateWords(SpvOp opcode,
   }
 }
 
+ir::Instruction* FoldInstructionInternal(
+    ir::Instruction* inst, std::function<uint32_t(uint32_t)> id_map,
+    InstructionBuilder<ir::IRContext::kAnalysisDefUse>* ir_builder,
+    bool* kill_inst) {
+  ir::IRContext* context = inst->context();
+  ir::Instruction* folded_inst = FoldInstructionToConstant(inst, id_map);
+  if (folded_inst != nullptr) {
+    if (*kill_inst) {
+      context->KillInst(inst);
+    }
+    return folded_inst;
+  }
+
+  SpvOp opcode = inst->opcode();
+  analysis::ConstantManager* const_manger = context->get_constant_mgr();
+
+  std::vector<uint32_t> ids;
+  std::vector<const analysis::Constant*> constants;
+  for (uint32_t i = 0; i < inst->NumInOperands(); i++) {
+    const ir::Operand* operand = &inst->GetInOperand(i);
+    if (operand->type != SPV_OPERAND_TYPE_ID) {
+      ids.push_back(0);
+      constants.push_back(nullptr);
+      continue;
+    }
+    ids.push_back(id_map(operand->words[0]));
+    const analysis::Constant* constant =
+        const_manger->FindDeclaredConstant(ids[i]);
+    constants.push_back(constant);
+  }
+
+  static FoldingRules rules;
+  for (FoldingRule rule : rules.GetRulesForOpcode(opcode)) {
+    bool new_inst = false;
+    folded_inst = rule(inst, ids, constants, ir_builder, &new_inst);
+    if (folded_inst != nullptr) {
+      if (*kill_inst) {
+        context->KillInst(inst);
```

Comment:
Maybe it would be better to leave the dead instructions in place and have a pass of adce clean them up.

Reply:
If we can avoid adding the instruction at all, then we won't have to kill.  The issue I have is that I do not want to create a lot of extra instructions.  Also, I prefer deleting instructions that we know are dead without any analysis.  Saves doing the analysis later.